hail/python/hailtop/batch/genetics/regenie/regenie.py

```diff
@@ -0,0 +1,310 @@
+    phenos_to_keep = {}
+    if step_args.phenoCol is not None and len(step_args.phenoCol):
+        for pheno in step_args.phenoCol:
+            phenos_to_keep[pheno] = True
+
+    if step_args.phenoColList is not None:
+        for pheno in step_args.phenoColList.split(","):
+            phenos_to_keep[pheno] = True
+
+    with hopen(step_args.phenoFile, "r") as f:
+        phenos = f.readline().strip().split(" ")[2:]
+
+    if not phenos_to_keep:
+        return phenos
+
+    phenos_final = []
+    for pheno in phenos:
+        if pheno in phenos_to_keep:
+            phenos_final.append(pheno)
+
+    return phenos_final
+
+
+def get_input(batch, step_args: argparse.Namespace):
+    add = {}
+    for name, val in vars(step_args).items():
+        if name in from_underscore:
+            name = from_underscore[name]
+
+        if name not in input_file_args or val is None:
+            continue
+
+        if name == "bed":
+            prefix = step_args.bed
+            add[name] = batch.read_input_group(
+                bed=f"{prefix}.bed", bim=f"{prefix}.bim", fam=f"{prefix}.fam")
+        elif name == "pgen":
+            prefix = step_args.pgen
+            add[name] = batch.read_input_group(
+                pgen=f"{prefix}.pgen", pvar=f"{prefix}.pvar", psam=f"{prefix}.psam")
+        else:
+            add[name] = batch.read_input(val)
+
+    return add
+
+
+def prepare_jobs(batch, args: BatchArgs, step1_args: argparse.Namespace, step2_args: argparse.Namespace):
+    j1 = batch.new_job(name='run-regenie')
+    j1.image('akotlar/regenie:9e7074f695e2b96bbb5af9d95f112d674c3260cd')
+    j1.cpu(args.cores)
+    j1.memory(args.memory)
+    j1.storage(args.storage)
+
+    in_step1 = get_input(batch, step1_args)
+
+    phenos = get_phenos(step1_args)
+    nphenos = len(phenos)
+
+    step1_output_prefix = step1_args.out
+    s1out = {"log": f"{step1_output_prefix}.log", "pred_list": f"{step1_output_prefix}_pred.list"}
+
+    for i in range(1, nphenos + 1):
+        s1out[f"{step1_output_prefix}_{i}"] = f"{step1_output_prefix}_{i}.loco"
+
+    if step1_args.lowmem:
+        for i in range(1, nphenos + 1):
+            pfile = f"{step1_args.lowmem_prefix}_l0_Y{i}"
+            s1out[pfile] = pfile
+
+    j1.declare_resource_group(**{step1_output_prefix: s1out})
+
+    cmd1 = []
+    for name, val in vars(step1_args).items():
+        if val is None or val is False or name == "step":
+            continue
+
+        if name in from_underscore:
+            name = from_underscore[name]
+
+        if name in input_file_args:
+            cmd1.append(f"--{name} {in_step1[name]}")
+        elif name == "out":
+            cmd1.append(f"--{name} {j1[step1_output_prefix]}")
+        elif isinstance(val, bool):
+            cmd1.append(f"--{name}")
+        elif name == "phenoCol":
+            for pheno in val:
+                cmd1.append(f"--{name} {pheno}")
+        else:
+            cmd1.append(f"--{name} {val}")
```

Comment:
> I just read your comments here: https://github.com/hail-is/hail/pull/9194/files/3719e7eef49811ec6149aa4dba0702b893ca89ca#r465142027.
> 
> 
> 
> Resource groups are for a set of files that have the same name but different file extensions. We don't have anything for a set of files that share a prefix but differ in idiosyncratic ways (e.g. end in "_3.loco").
> 
> 
> 
> Given my changes above, if you specify `--out gs://foo/bar/baz` then you'll get these files:
> 
> ```
> 
> gs://foo/bar/baz.log
> 
> gs://foo/bar/baz.pred_list
> 
> gs://foo/bar/baz.1_loco
> 
> gs://foo/bar/baz.2_loco
> 
> ...
> 
> ```
> 
> 
> 
> which, it seems violates your desire to preserve the file structure exactly.
> 
> 
> 
> I think in this circumstance, you'll have a better time using a folder rather than trying to use resource groups. For example, here I use a folder to hold two related files:
> 
> ```python3
> 
> In [17]: b = hb.Batch(backend=hb.ServiceBackend()) 
> 
>     ...: j = b.new_job() 
> 
>     ...: j.command(f'mkdir -p {j.out}') 
> 
>     ...: j.command(f'echo "Hello World" > {j.out}/hi') 
> 
>     ...: j.command(f'echo "Goodbye World" > {j.out}/bye') 
> 
>     ...: b.write_output(j.out, 'gs://danking/test123') 
> 
>     ...: b.run()                                                                                                                                                                                                                                     
> 
> ```
> 
> ```bash
> 
> # gsutil ls gs://danking/test123
> 
> gs://danking/test123/bye
> 
> gs://danking/test123/hi
> 
> ```
> 
> 
> 
> Something like this should work for regenie.
> 
> ```python3
> 
> j1.command(f'mkdir {j.output_dir}')
> 
> 
> 
> # ...
> 
>         elif name == "out":
> 
>             cmd1.append(f'--out {j.output_dir}/out')
> 
>             batch.write_output(j.output_dir, val)
> 
> ```
> 
> 
> 
> If I run this with `--out gs://danking/regenie-run123` I'll get:
> 
> ```
> 
> gs://danking/regenie-run123/out.log
> 
> gs://danking/regenie-run123/out_pred.list
> 
> gs://danking/regenie-run123/out_1.loco
> 
> ...
> 
> ```
> 
> Which seems fine?

> Resource groups are for a set of files that have the same name but different file extensions. We don't have anything for a set of files that share a prefix but differ in idiosyncratic ways (e.g. end in "_3.loco").

This doesn't make sense to me. It should be for a set of output files that are controlled by a single parameter. That is the case here: --out specifies the prefix of all our files. These files differ however by more than extension.

Reply:
> Which seems fine?

No, I want no out prefix. There is no way to tel regenie "don't output a file with a name", and batch always appends the identifier as an extension, which seems an unnecessary restriction.

Reply:
Also the output you described is not correct. I need {out_prefix}_Y1.regenie, not. {out_prefix}.Y1.regenie

Reply:
> I still don't really grok the input munging stuff, but the following seems somewhat easier to follow and eliminates the need for `get_input` and the preprocessing of the output files:
> 
> ```python
> 
>  def prepare_jobs(batch, args: BatchArgs, step1_args: argparse.Namespace, step2_args: argparse.Namespace):
> 
>     j1 = batch.new_job(name='run-regenie')
> 
>     # ...
> 
>     phenos = get_phenos(step1_args)
> 
>     nphenos = len(phenos)
> 
> 
> 
>     cmd1 = []
> 
>     for name, val in vars(step1_args).items():
> 
>         if name in from_underscore:
> 
>             name = from_underscore[name]
> 
> 
> 
>         if val is None or val is False:
> 
>             pass
> 
>         elif name == "bed":
> 
>             bed_files = batch.read_input_group(bed=f"{val}.bed", bim=f"{val}.bim", fam=f"{val}.fam")
> 
>             cmd1.append(f'--bed {bed_files}')
> 
>         elif name == "pgen":
> 
>             pgen_files = batch.read_input_group(pgen=f"{val}.pgen", pvar=f"{val}.pvar", psam=f"{val}.psam")
> 
>             cmd1.append(f'--pgen {pgen_files}')
> 
>         elif name in input_file_args:
> 
>             input_file = batch.read_input(val)
> 
>             cmd1.append(f'--{name} {input_file}')
> 
>         elif name == "out":
> 
>             j1.declare_resource_group(
> 
>                 output={'log': '{root}.log',
> 
>                         'pred_list': '{root}_pred.list',
> 
>                         **{f'{i}_loco': f'{{root}}_{i}.loco' for i in range(i, nphenos + 1)}}),
> 
>             cmd1.append(f'--out {j1.output}')
> 
>             batch.write_output(j1.output, val)
> 
>         elif name == "lowmem_prefix":
> 
>             j1.declare_resource_group(
> 
>                 lowmem={f'l0_Y{i}': f'{{root}}_l0_Y{i}'})
> 
>             cmd1.append(f'--lowmem_prefix {j1.lowmem}')
> 
>             batch.write_output(j1.lowmem, val)
> 
>         elif isinstance(val, bool):
> 
>             cmd1.append(f"--{name}")
> 
>         elif name == "phenoCol":
> 
>             for pheno in val:
> 
>                 cmd1.append(f"--phenoCol {pheno}")
> 
>         else:
> 
>             if name == "step":
> 
>                 assert val == "1"
> 
>             cmd1.append(f"--{name} {val}")
> 
> ```
> 
> 
> 
> I think I don't really understand the meaning of `--out FOO` as an argument to batch regenie. Presumably `FOO` is some GCS path to which we should copy step1's output? That's what I implemented above. Same for `--lowmem-prefix FOO`. Prefix doesn't really sound right though? It's just the remote location at which to store the lowmem files.

Batch --out is a remote location, yes. Regenie step 1/2 --out FOO is where regenie expects to write files, locally. Same for lowmem prefix. It's not a remote location (unless you mean remote as in something remote that emulates locally attached block storage). 

Note that regenie can run a single step, and users may want to get intermediate files. We can avoid duplicating interfaces by having the regenie basename named in step1/2 and the output directory provided in batch args. I had previously called this --outdir, switched because batch write_output is broken.

Reply:
> I still don't really grok the input munging stuff, but the following seems somewhat easier to follow and eliminates the need for `get_input` and the preprocessing of the output files

I don't see the benefit here. The input function creates the necessary resources (every argument it accepts is one that specifies a potential input file), then those resources are used as part of a short procedure to generate the string command.